# AoC Day 5

Jenna Jordan

5 December 2021

## Prompt

--- Day 5: Hydrothermal Venture ---

You come across a field of hydrothermal vents on the ocean floor! These vents constantly produce large, opaque clouds, so it would be best to avoid them if possible.

They tend to form in lines; the submarine helpfully produces a list of nearby lines of vents (your puzzle input) for you to review. For example:

```
0,9 -> 5,9
8,0 -> 0,8
9,4 -> 3,4
2,2 -> 2,1
7,0 -> 7,4
6,4 -> 2,0
0,9 -> 2,9
3,4 -> 1,4
0,0 -> 8,8
5,5 -> 8,2
```

Each line of vents is given as a line segment in the format x1,y1 -> x2,y2 where x1,y1 are the coordinates of one end the line segment and x2,y2 are the coordinates of the other end. These line segments include the points at both ends. In other words:

An entry like 1,1 -> 1,3 covers points 1,1, 1,2, and 1,3.
An entry like 9,7 -> 7,7 covers points 9,7, 8,7, and 7,7.
For now, only consider horizontal and vertical lines: lines where either x1 = x2 or y1 = y2.

So, the horizontal and vertical lines from the above list would produce the following diagram:

```
.......1..
..1....1..
..1....1..
.......1..
.112111211
..........
..........
..........
..........
222111....
```

In this diagram, the top left corner is 0,0 and the bottom right corner is 9,9. Each position is shown as the number of lines which cover that point or . if no line covers that point. The top-left pair of 1s, for example, comes from 2,2 -> 2,1; the very bottom row is formed by the overlapping lines 0,9 -> 5,9 and 0,9 -> 2,9.

To avoid the most dangerous areas, you need to determine the number of points where at least two lines overlap. In the above example, this is anywhere in the diagram with a 2 or larger - a total of 5 points.

Consider only horizontal and vertical lines. At how many points do at least two lines overlap?

In [1]:
import copy

In [2]:
# get input

with open("inputs/day5.txt") as file:
    inputs = file.read().splitlines()

#### create functions

In [3]:
def get_coordinates(inputs, drop_diagonals = True):
    coordinates = []
    for i in inputs:
        coords = i.split(" -> ")
        coords1 = tuple([int(x) for x in coords[0].split(",")])
        coords2 = tuple([int(x) for x in coords[1].split(",")])
        if drop_diagonals:
            x1, y1 = coords1
            x2, y2 = coords2
            if (x1 == x2) or (y1 == y2):
                coordinates.append((coords1, coords2))
        else:
            coordinates.append((coords1, coords2))
    return coordinates

In [4]:
def create_grid(x, y):
    grid = [[0] * x for i in range(y)]
    return grid

In [5]:
def mark_grid_horizontal_vertical(grid_input, coordinates):
    grid = copy.deepcopy(grid_input)
    for c in coordinates:
        (x1,y1),(x2,y2) = c
        # draw horizontal line
        if y1 == y2:
            if x2 > x1:
                for i in range(x1,x2+1):
                    grid[y1][i] += 1
            else:
                for i in range(x2,x1+1):
                    grid[y1][i] += 1  
        # draw vertical line
        elif x1 == x2:
            if y2 > y1:
                for i in range(y1,y2+1):
                    grid[i][x1] += 1
            else:
                for i in range(y2,y1+1):
                    grid[i][x1] += 1   
    return grid

In [6]:
def count_dangerous_areas(grid, threshold=2):
    danger_count = 0
    for row in grid:
        for spot in row:
            if spot >= threshold:
                danger_count += 1
    return danger_count

#### Part 1 Tests

In [7]:
test_inputs = ["0,9 -> 5,9", "8,0 -> 0,8", "9,4 -> 3,4", "2,2 -> 2,1", "7,0 -> 7,4", "6,4 -> 2,0", "0,9 -> 2,9", "3,4 -> 1,4", "0,0 -> 8,8", "5,5 -> 8,2"]

In [8]:
test_coordinates = get_coordinates(test_inputs, drop_diagonals = True)

In [9]:
test_grid = create_grid(10,10)

In [10]:
marked_test_grid = mark_grid_horizontal_vertical(test_grid, test_coordinates)
print(test_coordinates)
display(marked_test_grid)

[((0, 9), (5, 9)), ((9, 4), (3, 4)), ((2, 2), (2, 1)), ((7, 0), (7, 4)), ((0, 9), (2, 9)), ((3, 4), (1, 4))]


[[0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
 [0, 0, 1, 0, 0, 0, 0, 1, 0, 0],
 [0, 0, 1, 0, 0, 0, 0, 1, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
 [0, 1, 1, 2, 1, 1, 1, 2, 1, 1],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
 [2, 2, 2, 1, 1, 1, 0, 0, 0, 0]]

```
.......1..
..1....1..
..1....1..
.......1..
.112111211
..........
..........
..........
..........
222111....
```

In [11]:
count_dangerous_areas(marked_test_grid)

5

#### Part 1 with real inputs

In [12]:
coordinates_part1 = get_coordinates(inputs, drop_diagonals = True)

In [13]:
all_x = []
all_y = []
for coord in coordinates_part1:
    c1, c2 = coord
    x1, y1 = c1
    x2, y2 = c2
    all_x.append(x1)
    all_x.append(x2)
    all_y.append(y1)
    all_y.append(y2)
x_min = min(all_x)
x_max = max(all_x)
y_min = min(all_y)
y_max = max(all_y)
print(x_min, x_max)
print(y_min, y_max)
# just checking, but I want my grid to be 0 x 999

11 985
10 989


In [14]:
grid = create_grid(1000,1000)

In [15]:
marked_grid_part1 = mark_grid_horizontal_vertical(grid, coordinates_part1)

In [16]:
count_dangerous_areas(marked_grid_part1)

8111

### Part 1 Solution

8111 - CORRECT!

I made a very silly mistake at first and got an answer of 0 because I kept a test limiter of first 10 elements in a function... d'oh

--- Part Two ---

Unfortunately, considering only horizontal and vertical lines doesn't give you the full picture; you need to also consider diagonal lines.

Because of the limits of the hydrothermal vent mapping system, the lines in your list will only ever be horizontal, vertical, or a diagonal line at exactly 45 degrees. In other words:

An entry like 1,1 -> 3,3 covers points 1,1, 2,2, and 3,3.
An entry like 9,7 -> 7,9 covers points 9,7, 8,8, and 7,9.

```
0,9 -> 5,9
8,0 -> 0,8
9,4 -> 3,4
2,2 -> 2,1
7,0 -> 7,4
6,4 -> 2,0
0,9 -> 2,9
3,4 -> 1,4
0,0 -> 8,8
5,5 -> 8,2
```

Considering all lines from the above example would now produce the following diagram:

```
1.1....11.
.111...2..
..2.1.111.
...1.2.2..
.112313211
...1.2....
..1...1...
.1.....1..
1.......1.
222111....
```

You still need to determine the number of points where at least two lines overlap. In the above example, this is still anywhere in the diagram with a 2 or larger - now a total of 12 points.

Consider all of the lines. At how many points do at least two lines overlap?

In [17]:
def mark_grid_diagonal(grid_input, coordinates):
    grid = copy.deepcopy(grid_input)
    for c in coordinates:
        (x1,y1),(x2,y2) = c
        # draw diagonal line
        if (y1 != y2) and (x1 != x2):
            # if right down direction
            if (x2 > x1) and (y2 > y1):
                distance = (x2 - x1) + 1
                for i in range(distance):
                    grid[y1+i][x1+i] += 1
            # if left down direction
            elif (x1 > x2) and (y2 > y1):
                distance = (x1 - x2) + 1
                for i in range(distance):
                    grid[y1+i][x1-i] += 1
            # if right up direction
            elif (x2 > x1) and (y1 > y2):
                distance = (x2 - x1) + 1
                for i in range(distance):
                    grid[y1-i][x1+i] += 1
            # if left up direction
            elif (x1 > x2) and (y1 > y2):
                distance = (x1 - x2) + 1
                for i in range(distance):
                    grid[y1-i][x1-i] += 1
    return grid

#### part 2 tests

In [18]:
test_coordinates = get_coordinates(test_inputs, drop_diagonals = False)

In [19]:
marked_test_grid1 = mark_grid_horizontal_vertical(test_grid, test_coordinates)
marked_test_grid2 = mark_grid_diagonal(marked_test_grid1, test_coordinates)

In [20]:
display(marked_test_grid2)

[[1, 0, 1, 0, 0, 0, 0, 1, 1, 0],
 [0, 1, 1, 1, 0, 0, 0, 2, 0, 0],
 [0, 0, 2, 0, 1, 0, 1, 1, 1, 0],
 [0, 0, 0, 1, 0, 2, 0, 2, 0, 0],
 [0, 1, 1, 2, 3, 1, 3, 2, 1, 1],
 [0, 0, 0, 1, 0, 2, 0, 0, 0, 0],
 [0, 0, 1, 0, 0, 0, 1, 0, 0, 0],
 [0, 1, 0, 0, 0, 0, 0, 1, 0, 0],
 [1, 0, 0, 0, 0, 0, 0, 0, 1, 0],
 [2, 2, 2, 1, 1, 1, 0, 0, 0, 0]]


```
1.1....11.
.111...2..
..2.1.111.
...1.2.2..
.112313211
...1.2....
..1...1...
.1.....1..
1.......1.
222111....
```

### part 2 real thing

In [21]:
coordinates_part2 = get_coordinates(inputs, drop_diagonals = False)

In [22]:
marked_grid1_part2 = mark_grid_horizontal_vertical(grid, coordinates_part2)
marked_grid2_part2 = mark_grid_diagonal(marked_grid1_part2, coordinates_part2)

In [23]:
count_dangerous_areas(marked_grid2_part2)

22088

### Part 2 Solution

22088 - CORRECT